File: rgbserver.py
import os
import sys
import termios
import tty
import pigpio
import time
import threading
from flask import Flask, request, jsonify
from flask_cors import CORS, cross_origin

app = Flask(__name__)
CORS(app)

# Global Brightness
global bright
bright = 255

# Colors
global red
global green
global blue

# Pin numbers
RED_PIN = 17
GREEN_PIN = 22
BLUE_PIN = 24

global forceStop
forceStop = False

pi = pigpio.pi()


def updateColor(color, step):
    color += step

    if color > 255:
        return 255
    if color < 0:
        return 0

    return color


def setLights(pin, brightness):
    brightness = int(brightness)

    if brightness < 0:
        brightness = 0
    if brightness > 255:
        brightness = 255

    realBrightness = int(int(brightness) * (float(bright) / 255.0))
    
    pi.set_PWM_dutycycle(pin, realBrightness)

@app.route('/')
def homepage():
    return 'Pi RGB Server'

@app.route('/currentColor', methods=['GET'])
def currentColor():
    col = {
        'r': red,
        'g': green,
        'b': blue
    }

    return jsonify(col)

@app.route('/state', methods=['GET'])
def getState():
    col = {
        'r': red,
        'g': green,
        'b': blue
    }

    state = {
        'color': col,
        'brightness': bright
    }

    return jsonify(state)

@app.route('/brightness', methods=['POST'])
def setBrightness():
    global bright
    
    content = request.get_json()
    bright = content['brightness']

    if bright > 255:
        bright = 255
    if bright < 0:
        bright = 0

    setLights(RED_PIN, red)
    setLights(GREEN_PIN, green)
    setLights(BLUE_PIN, blue)

    return 'true'

@app.route('/fade', methods=['POST'])
def fadeColors():
    content = request.get_json()
    
    forceStop = True
    duration = content['duration']
    interval = content['interval']
    faderColors = content['colors']
    
    ctr = 0
    time.sleep(0.05)
    forceStop = False
    while(True):
        if forceStop:
            break
        
        color = faderColors[ctr % len(faderColors)]
        fadeColor(color, duration)

        ctr = ctr + 1
        time.sleep(interval)

    return 'true'

@app.route('/set', methods=['POST'])
def setColor():
    global red, green, blue
    
    content = request.get_json()

    red = content['r']
    green = content['g']
    blue = content['b']

    color = {
        'r': content['r'],
        'g': content['g'],
        'b': content['b']
    }

    #fadeColor(color, 250)

    setLights(RED_PIN, red)
    setLights(GREEN_PIN, green)
    setLights(BLUE_PIN, blue)

    return 'true'

def fadeColor(color, duration):
    global red, green, blue
    
    redInterval = (color['r'] - red) / float(duration)
    greenInterval = (color['g'] - green) / float(duration)
    blueInterval = (color['b'] - blue) / float(duration)

    print(redInterval, greenInterval, blueInterval)
 
    for i in range(0, duration):
        red += redInterval
        green += greenInterval
        blue += blueInterval

        setLights(RED_PIN, red)
        setLights(GREEN_PIN, green)
        setLights(BLUE_PIN, blue)

        time.sleep(0.001)

    red = int(red)
    green = int(green)
    blue = int(blue)


if __name__ == '__main__':
    red = 0
    green = 30
    blue = 30

    setLights(RED_PIN, red)
    setLights(GREEN_PIN, green)
    setLights(BLUE_PIN, blue)

    app.debug = True
    app.run(host='0.0.0.0', port=8008)

# pi.stop()

File: rgbserver.js
const Gpio = require('pigpio').Gpio;
const { Led, RGBLed } = require('pigpio-components');
const express = require('express');
const bodyParser = require('body-parser');
const cors = require('cors');
const app = express();
const http = require('http').Server(app);
const io = require('socket.io')(http);
const port = 8008;

const rgbLed = new RGBLed({ red: 17, green: 22, blue: 24 });
const redLed = new Led(17);
const greenLed = new Led(22);
const blueLed = new Led(24);

app.use(bodyParser.json())
app.use(cors());

/* Global Variables */
let currentColor = {
    r: 0,
    g: 255,
    b: 100
};

let globalBrightness = 255;

let fadeColorsInterval, fadeLightsInterval;

redLed.setValue(currentColor.r);
greenLed.setValue(currentColor.g);
blueLed.setValue(currentColor.b);

/* Socket.IO */
io.on('connection', socket => {
    console.log('User connected'); 
    socket.emit('updateColor', currentColor);

    socket.on('disconnect', () => {
        console.log('User disconnected');
    })
})

/* Get Methods */
app.get('/', (req, res) => res.send('RGB Server!'));

app.get('/currentColor', (req, res) => {
    roundColor(currentColor);
    res.json(currentColor)
})

app.get('/state', (req, res) => {
    roundColor(currentColor);

    const state = {
        color: currentColor,
        brightness: globalBrightness
    }

    res.json(state)
})

/* Post Methods */
app.post('/brightness', (req, res) => {
    
    globalBrightness = Math.round(req.body.brightness);

    setLights("red", currentColor.r);
    setLights("green", currentColor.g);
    setLights("blue", currentColor.b);

    res.json(true)
})

app.post('/set', (req, res) => {
    killIntervals();

    currentColor = {
        r: req.body.r,
        g: req.body.g,
        b: req.body.b
    };

    setLights("red", currentColor.r);
    setLights("green", currentColor.g);
    setLights("blue", currentColor.b);

    res.json(true);
})

app.post('/fade', (req, res) => {
    fadeLights(req.body.color, req.body.duration);

    res.json(true);
})

app.post('/fadeColors', (req, res) => {
    if(fadeColorsInterval) {
        clearInterval(fadeColorsInterval);
    }

    let faderColors = req.body.colors;
    let fcLength = faderColors.length;
    let ctr = 0;

    fadeLights(faderColors[ctr++ % fcLength], req.body.duration);
   
    fadeColorsInterval = setInterval(() => {
        if(faderColors.length === 0) {
            clearInterval(fadeColorsInterval);
            return;
        }
        fadeLights(faderColors[ctr++ % fcLength], req.body.duration);
    }, req.body.interval * 1000)
    
    res.json(true);
})

/* Server Subroutines */
function setLights(color, brightness) {
    let realBrightness = brightness * (globalBrightness / 255);
    if(realBrightness > 255) realBrightness = 255;
    if(realBrightness < 0) realBrightness = 0;
    
    realBrightness = Math.round(realBrightness);
    brightness = Math.round(brightness);

    if(color == "red") {
        redLed.setValue(realBrightness)
    }

    if(color == "green") {
        greenLed.setValue(realBrightness)
    }

    if(color == "blue") {
        blueLed.setValue(realBrightness)
    }
    
    io.sockets.emit('updateColor', currentColor);
}

function fadeLights(destColor, duration) {
    // Duration in ms
    let redStep = (destColor.r - currentColor.r) / duration;
    let greenStep = (destColor.g - currentColor.g) / duration;
    let blueStep = (destColor.b - currentColor.b) / duration;

    if(fadeLightsInterval) {
        clearInterval(fadeLightsInterval);
    }

    let fadeInterval = null;
    
    let ctr = 0;

    fadeLightsInterval = setInterval(() => {
        if(ctr >= duration) {
            clearInterval(fadeInterval);
            return;
        }

        currentColor.r += redStep;
        currentColor.g += greenStep;
        currentColor.b += blueStep;
        
        setLights("red", currentColor.r);
        setLights("green", currentColor.g);
        setLights("blue", currentColor.b);

        ++ctr;
    }, 1)

}

function killIntervals() {
    console.log('Killing intervals');

    if(fadeLightsInterval) {
        clearInterval(fadeLightsInterval);
    }

    if(fadeColorsInterval) {
        clearInterval(fadeColorsInterval);
    }
}

function roundColor(color) {
    color.r = Math.round(color.r);
    color.g = Math.round(color.g);
    color.b = Math.round(color.b);
}

http.listen(port, () => console.log(`Server listening on port ${port}`));
